sputnikdao-factory2/src/lib.rs

```diff
@@ -138,6 +143,108 @@ impl SputnikDAOFactory {
+
+        // Compute and use the correct amount needed for storage
+        let blob_len = dao_contract_code.len();
+        let storage_cost = ((blob_len + 32) as u128) * env::storage_byte_cost();
+
+        // Confirm payment before proceeding
+        assert!(
+            storage_cost <= env::attached_deposit(),
+            "Must at least deposit {} to store",
+            storage_cost
+        );
+
+        // refund the extra cost
+        let extra_attached_deposit = env::attached_deposit() - storage_cost;
+        Promise::new(account_id.clone()).transfer(extra_attached_deposit);
+
+        // Create a promise toward given account.
+        let promise_id = env::promise_batch_create(&account_id);
+        env::promise_batch_action_function_call(
+            promise_id,
+            method_name,
+            &dao_contract_code,
+            storage_cost,
+            env::prepaid_gas() - env::used_gas() - GAS_STORE_CONTRACT_LEFTOVER,
+        );
+        env::promise_return(promise_id);
+    }
+
+    /// Allows a DAO to remove the blob stored in its DAO storage, and reclaim the storage cost
+    pub fn remove_contract_self(&mut self, code_hash: Base58CryptoHash) {
+        let account_id = env::predecessor_account_id();
+        let factory_id = env::current_account_id();
+        let method_name = "remove_blob";
+
+        // NOTE: Not verifing the hash, in case factory removes a hash before DAO does
+        let method_args = &json!({ "hash": &code_hash }).to_string().into_bytes();
+        let callback_method = "on_remove_contract_self";
+        let callback_args = &json!({ "account_id": &account_id, "code_hash": &code_hash }).to_string().into_bytes();
+
+        // Create a promise toward given account.
+        let promise_id = env::promise_batch_create(&account_id);
+        env::promise_batch_action_function_call(
+            promise_id,
+            method_name,
+            method_args,
+            NO_DEPOSIT,
+            GAS_STORE_CONTRACT_LEFTOVER,
+        );
+        // attach callback to the factory.
+        let _ = env::promise_then(
+            promise_id,
+            factory_id,
+            callback_method,
+            callback_args,
+            NO_DEPOSIT,
+            ON_REMOVE_CONTRACT_GAS,
+        );
+        env::promise_return(promise_id);
+    }
+
+    /// Upon blob remove, compute the balance (if any) that got paid to the factory,
+    /// since it was the "owner" of the blob stored on the DAO.
+    /// Send this balance back to the DAO, since it was the original funder
+    #[private]
+    pub fn on_remove_contract_self(&mut self, account_id: AccountId, code_hash: Base58CryptoHash) -> bool {
+        if near_sdk::is_promise_success() {
+            // Compute the actual storage cost for an accurate refund
+            let hash: CryptoHash = code_hash.into();
+            let dao_contract_code = env::storage_read(&hash).expect("CODE_HASH_NONEXIST");
```

Comment:
what if the `hash` is not in the storage anymore because it was removed by the factory? The user will not get refunded and the money will stay in the factory.

Reply:
Yeah, i thought about that as well. I realize thats a possibility, but I would _hope_ standard practice of factory will not remove versions until all DAOs confirmed upgraded beyond. 
I just couldnt find a way yet that helps users retrieve funds in that case, and is safe.

Reply:
Anyway, this entire flow is used only for v2 -> v3 upgrade. The idea that the V3 code will be wiped out from the factory is very unlikely so we should be fine.